google-cloud-spanner/src/test/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClientMockServerTest.java

```diff
@@ -467,6 +480,239 @@ public void testWriteAtLeastOnceWithExcludeTxnFromChangeStreams() {
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsAcquired().get());
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsReleased().get());
+  }
+
+  @Test
+  public void testReadWriteTransactionUsingAsyncTransactionManager() throws Exception {
+    // Updates executed within a R/W transaction via AsyncTransactionManager should use a
+    // multiplexed session.
+    // During a retry (due to an ABORTED error), the transaction should use the same multiplexed
+    // session as before, assuming the maintainer hasn't run in the meantime.
+    final AtomicInteger attempt = new AtomicInteger();
+    CountDownLatch abortedLatch = new CountDownLatch(1);
+    DatabaseClientImpl client =
+        (DatabaseClientImpl) spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+    try (AsyncTransactionManager manager = client.transactionManagerAsync()) {
+      TransactionContextFuture transactionContextFuture = manager.beginAsync();
+      while (true) {
+        try {
+          attempt.incrementAndGet();
+          AsyncTransactionStep<Void, Long> updateCount =
+              transactionContextFuture.then(
+                  (transaction, ignored) -> transaction.executeUpdateAsync(UPDATE_STATEMENT),
+                  MoreExecutors.directExecutor());
+          updateCount.then(
+              (transaction, ignored) -> {
+                if (attempt.get() == 1) {
+                  mockSpanner.abortTransaction(transaction);
+                  abortedLatch.countDown();
+                }
+                return ApiFutures.immediateFuture(null);
+              },
+              MoreExecutors.directExecutor());
+          abortedLatch.await(10L, TimeUnit.SECONDS);
+          CommitTimestampFuture commitTimestamp = updateCount.commitAsync();
+          assertThat(updateCount.get()).isEqualTo(UPDATE_COUNT);
+          assertThat(commitTimestamp.get()).isNotNull();
+          assertThat(attempt.get()).isEqualTo(2);
+          break;
+        } catch (AbortedException e) {
+          transactionContextFuture = manager.resetForRetryAsync();
+        }
+      }
+    }
+
+    List<ExecuteSqlRequest> executeSqlRequests =
+        mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(2, executeSqlRequests.size());
+    assertEquals(executeSqlRequests.get(0).getSession(), executeSqlRequests.get(1).getSession());
+
+    // Verify the requests are executed using multiplexed sessions
+    for (ExecuteSqlRequest request : executeSqlRequests) {
+      assertTrue(mockSpanner.getSession(request.getSession()).getMultiplexed());
+    }
+
+    assertNotNull(client.multiplexedSessionDatabaseClient);
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsAcquired().get());
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsReleased().get());
+  }
+
+  @Test
+  public void testReadWriteTransactionUsingAsyncRunner() throws Exception {
```

Comment:
Added test for AsyncRunner to verify the non-blocking behavior and verified that it works with multiplexed sessions.
However, the same test will not work for AsyncTransactionManager since it is blocking and hence I have not added it in the test suite.

Reply:
Ack